auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,196 @@
+
+        /// Checks for correct asset_id being sent and high enough amount being sent
+        require(msg_asset_id() == storage.asset_id, Error::WrongAssetSent);
+        require(msg_amount() >= calculate_price(auction_id), Error::BidTooLow);
+
+        /// Cannot bid before auction starts
+        require(height() >= auction.start_time, Error::AuctionNotYetStarted);
+        
+        /// If ended == true, someone already bid or the admin prematurely ended the auction
+        require(!auction.ended, Error::AuctionAlreadyEnded);
+
+        /// Disallows furthur bids
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+
+        /// If someone sends more than the current price, refund the extra amount 
+        if msg_amount() > calculate_price(auction_id) {
+            let return_amount = msg_amount() - calculate_price(auction_id);
+            transfer_to_output(return_amount, storage.asset_id, get_sender());
+        }
+
+        /// Logic on win put into the win function. Using a function here so that its easier to modify the logic 
+        /// post-auction-win and so this contract template can be reused for multiple things
+        win(auction_id);
+    }
+
+    fn setup_auction(opening_price: u64, reserve_price: u64, start_time: u64, end_time: u64) -> u64 {
+        require(storage.constructed == true, Error::ContractNotConstructedYet);
+
+        require(get_sender() == storage.admin, Error::SenderNotAdmin);
+        require(opening_price > reserve_price, Error::EndPriceCannotBeLargerThanStartPrice);
+        require(end_time > height(), Error::AuctionCannotEndInThePast);
+        require(start_time >= height(), Error::AuctionCannotStartInThePast);
+        require(end_time > start_time, Error::AuctionCannotEndBeforeItStarts);
+
+        storage.latest_auction_id = storage.latest_auction_id + 1;
+        let current_auction_id = storage.latest_auction_id;
+
+        let mut auction = storage.auctions.get(current_auction_id);
+        auction.opening_price = opening_price;
+        auction.reserve_price = reserve_price;
+        auction.start_time = start_time;
+        auction.end_time = end_time;
+        auction.ended = false;
+        
+        storage.auctions.insert(current_auction_id, auction);
+
+        /// Returns the auction being setup's auction_id
+        current_auction_id
+    }
+
+    fn end_auction(auction_id: u64) {
+        require(storage.constructed == true, Error::ContractNotConstructedYet);
+
+        /// Only the admin can end the auction (prematurely)
+        require(get_sender() == storage.admin, Error::SenderNotAdmin);
+
+        /// If there is no auction going on currently the ended value will already be true so no need to check for that case
+        let mut auction = storage.auctions.get(auction_id);
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+    }
+
+}
+
+fn win(auction_id: u64) {
```

Comment:
add documentation stating that this is a placeholder that is meant to have additional logic added if the user wants to